p2p/conn/connection.go

```diff
@@ -446,15 +476,16 @@ FOR_LOOP:
 		switch pktType {
 		case packetTypePing:
 			// TODO: prevent abuse, as they cause flush()'s.
+			// https://github.com/tendermint/tendermint/issues/1190
 			c.Logger.Debug("Receive Ping")
 			select {
 			case c.pong <- struct{}{}:
 			case <-c.quit:
 				break FOR_LOOP
 			}
 		case packetTypePong:
-			// do nothing
 			c.Logger.Debug("Receive Pong")
+			c.stopPongTimer()
```

Comment:
if we receive ping and pong at the same time, wont this race with the `c.pongTimer = ...` in the sendRoutine ? Maybe we need a test where that happens for the race detector to pick it up

Reply:
Oh the race detector actually detects this for me

Reply:
I guess we could just wrap pongTimer in a mutex but that doesn't really feel right.

An alternative is maybe to use a separate go-routine to handle a pongTimer. Then sendRoutine and recvRoutine can just talk to the pongTimerRoutine via channel/s when they need to start or stop the pongTimer.

Reply:
What if we use the same pongTimeoutCh to signal to cancel timeout? See my last commit.